src/components/virtual-gamepad-controls.js

```diff
@@ -0,0 +1,188 @@
+
+    this.onEnterVr = this.onEnterVr.bind(this);
+    this.onExitVr = this.onExitVr.bind(this);
+    this.el.sceneEl.addEventListener("enter-vr", this.onEnterVr);
+    this.el.sceneEl.addEventListener("exit-vr", this.onExitVr);
+  },
+
+  onJoystickChanged(event, data) {
+    if (event.target.id === this.leftStick.id) {
+      if (event.type === "move") {
+        // Set velocity vector on left stick move
+        var angle = data.angle.radian;
+        var force = data.force < 1 ? data.force : 1;
+        var x = Math.cos(angle) * force;
+        var z = Math.sin(angle) * -force;
+        this.velocity.set(x, 0, z);
+      } else {
+        this.velocity.set(0, 0, 0);
+      }
+    } else {
+      if (event.type === "move") {
+        // Set pitch and yaw angles on right stick move
+        var angle = data.angle.radian;
+        var force = data.force < 1 ? data.force : 1;
+        this.yaw = Math.cos(angle) * -force;
+        // this.pitch = Math.sin(angle) * force;
+      } else {
+        this.yaw = 0;
+        // this.pitch = 0;
+      }
+    }
+  },
+
+  tick(t, dt) {
+    const deltaSeconds = dt / 1000;
+    const lookSpeed = THREE.Math.DEG2RAD * this.data.lookSpeed * deltaSeconds;
+    const obj = this.el.object3D;
+    const pivot = this.cameraEl.object3D;
+    const distance = this.data.movementSpeed * deltaSeconds;
+
+    this.pivotPos.copy(pivot.position);
+    this.pivotPos.applyMatrix4(obj.matrix);
+    this.trans.setPosition(this.pivotPos);
+    this.transInv.makeTranslation(
+      -this.pivotPos.x,
+      -this.pivotPos.y,
+      -this.pivotPos.z
+    );
+    this.rotationMatrix.makeRotationAxis(this.rotationAxis, obj.rotation.y);
+    this.rotationInvMatrix.makeRotationAxis(this.rotationAxis, -obj.rotation.y);
+    this.camRotationMatrix.makeRotationAxis(
+      this.rotationAxis,
+      pivot.rotation.y
+    );
+    this.camRotationInvMatrix.makeRotationAxis(
+      this.rotationAxis,
+      -pivot.rotation.y
+    );
+    this.move.makeTranslation(
+      this.velocity.x * distance,
+      this.velocity.y * distance,
+      this.velocity.z * distance
+    );
+
+    this.yawMatrix.makeRotationAxis(this.rotationAxis, lookSpeed * this.yaw);
+
+    // Translate to middle of playspace (player rig)
+    obj.applyMatrix(this.transInv);
+    // Zero playspace (player rig) rotation
+    obj.applyMatrix(this.rotationInvMatrix);
+    // Zero camera (head) rotation
+    obj.applyMatrix(this.camRotationInvMatrix);
+    // Apply joystick translation
+    obj.applyMatrix(this.move);
+    // Apply joystick yaw rotation
+    obj.applyMatrix(this.yawMatrix);
+    // Reapply camera (head) rotation
+    obj.applyMatrix(this.camRotationMatrix);
+    // Reapply playspace (player rig) rotation
+    obj.applyMatrix(this.rotationMatrix);
+    // Reapply playspace (player rig) translation
+    obj.applyMatrix(this.trans);
+
+    // @TODO this is really ugly, can't just set the position/rotation directly or they wont network
```

Comment:
How is it ugly?

Reply:
Because you would want to just be able to call `this.el.setAttribute("rotation", someQuaternion)` and not have to create a new object with components all in degrees only to have aframe convert them back to radians and stick them back in a quaternion.

Reply:
Sure, but that's a-frame. I don't see the value of pointing out in a TODO if we can't actually do anything about it.